src/Deserialization.cpp

```diff
@@ -0,0 +1,1325 @@
+#include "Deserialization.h"
+#include <fstream>
+#include <functional>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "FindCalls.h"
+#include "Func.h"
+#include "Function.h"
+#include "IR.h"
+#include "Schedule.h"
+#include "halide_ir_generated.h"
+
+namespace Halide {
+namespace Internal {
+class Deserializer {
+public:
+    Deserializer() = default;
+
+    explicit Deserializer(const std::map<std::string, Parameter> &external_params)
+        : external_params(external_params) {
+    }
+
+    Pipeline deserialize(const std::string &filename);
+
+    Pipeline deserialize(std::istream &in);
+
+private:
+    // Helper function to deserialize a homogenous vector from a flatbuffer vector,
+    // does not apply to union types like Stmt and Expr or enum types like MemoryType
+    template<typename src, typename dst>
+    std::vector<dst> deserialize_vector(const flatbuffers::Vector<::flatbuffers::Offset<src>> *flatbuffer_vec, std::function<dst(Deserializer &, const src *)> deserialize_func) {
+        user_assert(flatbuffer_vec != nullptr) << "deserializing a null vector\n";
```

Comment:
I think the lines are kind of blurry: We can call it a user-error if the user provided broken/not-compatible file input, otherwise we can call it an internal error

Reply:
But in both ways we should stay consistent: pick one from `user_error`/`internal_error` and use across serialization implementations.